src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -626,6 +625,56 @@ public byte[] Http2AltSvcOriginUri
             return await GetHttpConnectionAsync(request, cancellationToken).ConfigureAwait(false);
         }
 
+        private Http2Connection? GetHttp2ConnectionAcceptingNewStreams()
+        {
+            IReadOnlyList<Http2Connection>? localConnections = _http2Connections;
+
+            if (localConnections == null)
+            {
+                return null;
+            }
+
+            for (int i = 0; i < localConnections.Count; i++)
+            {
+                Http2Connection http2Connection = localConnections[i];
+
+                if (!_poolManager.Settings.EnableMultipleHttp2Connections || http2Connection.CanAddNewStream)
```

Comment:
Previously, we were recycling the connection everytime its lifetime expired. Why is it now different (only when `CanAddNewStream`)? I'm sure I'm missing something, please help me understand.

Reply:
I have just changed this logic and it now checks expiration before trying to acquire a stream slot. Please, check the latest version.